### PR TITLE
Fix/353 get file content sec access

### PR DESCRIPTION
**Pull request by ianliuwk1019**
This PR is for #353. The fixes include:
- Use `GUARD_OPTIONS.ANONYMOUS_LIMITED` security for Attachment.controller.ts `getFileContents` endpoint.
- Add swagger @ApiProduces('application/octet-stream') for response type.
- Fix bug in `isViewAuthorized` due to Attachment entity not returning attachmentTypeCode.
- Generate new api-client for frontend.
- Use convenient lib `file-saver` for handling downloaded file for public/admin frontend.
- Fix previous download attachment using <a> direct link by replacing it with (click) event and call api backend with proper user header.
- Adjust backend attachment service `findFileDatabase` entity loading options based on discussion.
- Fix Interaction attachment update issue from calling TypeORM repository.update() function due to TypeORM behavior change.

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-381.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-381.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-381.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-381.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-381.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-381.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

### CODE REVIEW

**Comment by basilv** [comment]
When testing, it looks like the attachment can now only be downloaded, not viewed directly in the browser... I can't recall if this is a change in behavior, but it seems okay to me.

File: api/src/app/modules/attachment/attachment.service.ts [hunk 1]
```diff
@@ -92,14 +92,15 @@ export class AttachmentService extends DataService<Attachment, Repository<Attach
   }
 
   async isDeleteAuthorized(entity: Attachment, user?: User):Promise<boolean> {
+    const attachmentTypeCode = entity.attachmentType?.code;
```
Comment (basilv) [inline api/src/app/modules/attachment/attachment.service.ts:95]: I don't understand why this change was necessary, maybe it was broken before. AttachmentCreateRequest DTO has attachmentTypeCode, but the AttachmentResponse DTO has attachmentType. The entity has both attachmentTypeCode (the join column) and attachmentType...

File: api/src/app/modules/attachment/attachment.service.ts [hunk 2]
```diff
@@ -110,8 +111,8 @@ export class AttachmentService extends DataService<Attachment, Repository<Attach
   }
 
   async isViewAuthorized(entity: Attachment, user?: User): Promise<boolean> {
-
-    if (entity.attachmentTypeCode == AttachmentTypeEnum.PUBLIC_NOTICE || entity.attachmentTypeCode == AttachmentTypeEnum.SUPPORTING_DOC) {
+    const attachmentTypeCode = entity.attachmentType?.code;
+    if (attachmentTypeCode == AttachmentTypeEnum.PUBLIC_NOTICE || attachmentTypeCode == AttachmentTypeEnum.SUPPORTING_DOC) {
       // These document types are viewable by the public.
```
Comment (basilv) [inline api/src/app/modules/attachment/attachment.service.ts:116]: I don't remember if supporting documents should be visible to the public. Interactions should definitely NOT be.

File: api/src/app/modules/attachment/attachment.service.ts
Comment (basilv) [inline api/src/app/modules/attachment/attachment.service.ts:1]: getFileContents() I believe needs a security check with the same logic as isViewAuthorized.

File: api/src/app/modules/attachment/attachment.service.ts (hunk 1 above)
Comment (basilv) [reply api/src/app/modules/attachment/attachment.service.ts:95]: I suspect if the entity was loaded but not its relations, attachmentType can be null, but attachmentTypeCode should never be null?

File: api/src/app/modules/attachment/attachment.service.ts (hunk 2 above)
Comment (basilv) [reply api/src/app/modules/attachment/attachment.service.ts:116]: The pre-existing code had this same logic (public can see supporting documents) so lets keep it that way.

**Comment by basilv** [comment]
I got an error trying to add an attachment to an existing interaction - see screenshot 
image

**Review by basilv — CHANGES_REQUESTED** [review]
Error found in testing, not sure about the attachmentTypeCode change, haven't tested public component.

**Comment by ianliuwk1019** [comment]
> I got an error trying to add an attachment to an existing interaction - see screenshot image

Strange, needs to investigate this one.
Adding attachment (new) is good, update seems to be a problem but returning 503, strange.

================================

Looks like it is a different issue than this ticket and it is a new one we didn't discovered before.
Glad you found it @basilv . The issue seems, again, behavior change at **TypeORM**.
Attach images for before-fix and after-fix:

image

image

The `.updateEntity(id, {attachmentId: undefined}, entity)` isn't working anymore for the argument **undefined** (which was working before). It has to be `.updateEntity(id, {attachmentId: null}, entity)` **null** for the argument. Then the partial update will work, so subsequent old attachment can be deleted and a new attachment to the same Interaction entity can be added.

File: api/src/app/modules/attachment/attachment.service.ts
Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts:1]: The old code "findFileDatabase()" function was called by getFileContents() which does check for isViewAuthorized().

File: api/src/app/modules/attachment/attachment.service.ts (hunk 2 above)
Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts:116]: Yes, the business logic for condition did not get changed. What I was trying to do is to use the same constant for value of 'attachmentTypeCode'. However, I do change it to get value from the AttachmentTypeCode relation entity that's available from Attachment entity since the other `@RelationId` is for some reason empty that is causing the checking fails. I do document this point in ticket.

File: api/src/app/modules/attachment/attachment.service.ts (hunk 1 above)
Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts:95]: Actually, while I was testing, the Attachment entity got back has `attachmentType` relation loaded but not the `attachmentTypeCode`. This is strange I found and documented in ticket. So I search the code where has this logic to indirectly use the loaded `attachmentType` to get the `code` so it does not break.

**Comment by ianliuwk1019** [comment]
> When testing, it looks like the attachment can now only be downloaded, not viewed directly in the browser... I can't recall if this is a change in behavior, but it seems okay to me.

I would feel this is okay, also I don't recall we have that 'view' before. The 'summary' has but it is different (by css), not downloadin. I can try revert back tp previous <a> link with backend open (just temporarily) and see what was the behavior before.

**Comment by basilv** [comment]
Issues with line 167 addCommonRelationsToFindOptions in findFileDatabase in attachment.service.ts

**Comment by ianliuwk1019** [comment]
> Issues with line 167 addCommonRelationsToFindOptions in findFileDatabase in attachment.service.ts

This duplicated call is removed. However, it does not fix entity loading issue, behave the same.

File: api/src/app/modules/attachment/attachment.service.ts (hunk 1 above)
Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts:95]: The duplicated call on line 167 addCommonRelationsToFindOptions in findFileDatabase is removed but does not fix the loading of `attachmentTypeCode`.
However, during removing, I found the list of "options" actually didn't state "attachmentTypeCode". If I add this `attachmentTypeCode` into the list, it shows up.

But still I chose not to add this `attachmentTypeCode`, it makes sense to me to leave it out since we already specify `attachmentType` relation entity to load so the code is already there as `[ 'id', 'projectId', 'fileName', 'attachmentType']`

File: api/src/app/modules/attachment/attachment.service.ts [hunk 3]
```diff
@@ -164,8 +164,9 @@ export class AttachmentService extends DataService<Attachment, Repository<Attach
     // NestJS/TypeORM has breaking change for find API which only accespts 'option' argument and
     // no 'id' for findOne(id, options) like before. We need to specifically build the 'options.where'
     // for the TypeORM api now.
-    const revisedOptions = this.addCommonRelationsToFindOptions(this.addCommonRelationsToFindOptions(
-        { select: [ 'id', 'projectId', 'fileName', 'attachmentType' ] }));
+    const revisedOptions = this.addCommonRelationsToFindOptions(
+        { select: [ 'id', 'projectId', 'fileName', 'attachmentType']}
```
Comment (basilv) [inline api/src/app/modules/attachment/attachment.service.ts]: This logic is still weird. I think we should just be calling this.findEntityWithCommonRelations(id) instead of all of this logic (up to line 171) which is duplicating (incorrectly) the generic logic in data.service. The existing logic is specifying what columns to return, so that's why attachmentTypeCode is null as it isn't included in the list. 

I'm trying to assess if there's a reason we did the logic special this way, but I can't think of one at the moment.

Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts]: You are correct @basilv . I wasn't thinking that much when I did the adjustment for TypeORM upgrade on passing `options` to TypeORM (but not the logic change than in the past). I don't know why previous logic needs to specify columns, and I don't see a condition to return different columns depending on the 'user' argument to return, but I replace it with `const entity:Attachment = await this.findEntityWithCommonRelations(id)` and it is still working from public/admin.

Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts]: Ah~, I probably understand now with the previous code. It seems it is trying to **reduce** "**select**" return from repository.
So, other functions like "**getFileContent**" that uses "**this.findFileDatabase(id, user)**" function only needs these fields' value from **Attachment Entity**:

> **id,
> projectId,
> fileName**

plus, it also needs "**attachmentType**"(entity) or "**attachmentTypeCode**"(just code) (either one) to do validation on `this.isViewAuthorized`.
See the calling function:
```
async getFileContent(id: number, user?: User): Promise<AttachmentFileResponse> {
    const attachmentFileResponse = await this.findFileDatabase(id, user);

    //Creating the objectName for the Object Storage
    const objectName = this.createObjectUrl(attachmentFileResponse.projectId, attachmentFileResponse.id, attachmentFileResponse.fileName )
```
That's probably why it needs a list of **["id", "projectId", "fileName", "attachmentType"]**

Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts]: But on the other hand, the attachment.service.ts has 
```  
protected getCommonRelations(): string[] {
    return ['attachmentType'];
  }
```
So basically, they are equivalent. Although "**project**" entity isn't listed as common relation in both of them( through `this.findEntityWithCommonRelations(id)` or through specific selection `["id", "projectId", "fileName", "attachmentType"]`), "**project**" entity is always returned (as "eager" fetch I guess).
I will change to use `this.findEntityWithCommonRelations(id)`.

File: api/src/app/modules/attachment/attachment.service.ts (hunk 1 above)
Comment (ianliuwk1019) [reply api/src/app/modules/attachment/attachment.service.ts:95]: Replaced old code to use `this.findEntityWithCommonRelations(id)` now.

**Review by basilv — APPROVED** [review]
